[16] [ADD] sale_import_base

florian-dacosta opened the pull request:
Add module that create a sale order from a payload. There wil be another module on top of this one that exposes a REST api, to allow external system to send sale orders to Odoo.

It has been refactore and aims to replace https://github.com/OCA/sale-channel/pull/32 and include https://github.com/OCA/queue/pull/737 following comment here https://github.com/OCA/queue/pull/737#issuecomment-2639294595
and  here https://github.com/OCA/sale-channel/pull/32#issuecomment-2647435642

@clementmbr @hparfr @sebastienbeau

kevinkhao committed:
[ADD] sale_import_base, sale_import_delivery_carrier, sale_import_rest

kevinkhao committed:
[FIX] Remove batch run test

kevinkhao committed:
[IMP] rename to respect OCA variable

kevinkhao committed:
[IMP] Improve tests

kevinkhao committed:
[IMP] Akretion copyrights

kevinkhao committed:
[IMP] remove _si prefixes on processor component

kevinkhao committed:
[IMP] rename json_import to data

kevinkhao committed:
[REF] Simplify sale importer component, cleanup tests

kevinkhao committed:
[FIX] Inheritance for datamodels, delivery carrier logic and tests

kevinkhao committed:
[IMP] Create sale order before sale order lines due to onchange invalidating cache
Test improvements on delivery carrier line generation

kevinkhao committed:
Update sale_import_base/components/importer_sale_channel.py

Co-authored-by: beau sebastien <sebastien.beau@akretion.com>

kevinkhao committed:
Update sale_import_base/tests/test_sale_order_import.py

Co-authored-by: beau sebastien <sebastien.beau@akretion.com>

kevinkhao committed:
Add tests, improve code logic, move validator from datamodel to importer

kevinkhao committed:
Tweaks to invoice, add force number and force date to sale order

sebastienbeau committed:
[IMP] simplify the code

- add a specific validator for customer (as required field are not the
same)
- remove currency as it's deduce from pricelist
- remove useless validate_misc due to previous change

sebastienbeau committed:
[REF] rename transaction_id to acquirer_reference and map it correctly

sebastienbeau committed:
[FIX] acquirer_reference should not be required (check, cash...)

sebastienbeau committed:
[IMP] we should pass the name of the sale order

sebastienbeau committed:
[IMP] make external amount optional as we may no check them

sebastienbeau committed:
[IMP] add possibility to set the phone and mobile

sebastienbeau committed:
[IMP] make the pricelist optionnal by reading it from the channel

sebastienbeau committed:
[FIX] add missing partner on transaction

kevinkhao committed:
[IMP] Modifications related to changing "required" attribute on datamodels: includes SO generation, tests, components, demo data
Add tests for cases where data content is full/minimal/mixed
Fix state_code: search also according to country_id
Fix state_code, country_code validation and search

kevinkhao committed:
[IMP] Add date to sale order

kevinkhao committed:
[FIX] Revert required attr on amounts

kevinkhao committed:
[REF] On import, remove pre-emptive validation and error list, instead raise on first error
[IMP] Validate payment currency as it should match Sale Order's pricelist's currency

kevinkhao committed:
[IMP] Improve UI for sale channel form view

kevinkhao committed:
[14.0][MIG] Migration to 14.0

kevinkhao committed:
[IMP] run pre-commit

kevinkhao committed:
[14.0] 14.0 Branch: Initial commit with no installable modules

sebastienbeau committed:
[IMP] add the possibility to confirm order and generate invoice

kevinkhao committed:
[IMP] Improve pluggability of ImporterSaleChannel component

kevinkhao committed:
[FIX] Fix partner versioning using alternate way of dealing with addresses

kevinkhao committed:
[IMP] Add types to addresses

kevinkhao committed:
[IMP] Create addresses in inactive state, add tests

kevinkhao committed:
[IMP] Add notion of choosing either external identifier or Odoo native numbering
Minor typo fix
Minor test refactor

sebastienbeau committed:
[IMP] improve error message

sebastienbeau committed:
[FIX] fix missing date

florian-dacosta committed:
[FIX] Avoid sale order duplicates in case a same chunk is played mutliple times

florian-dacosta committed:
[FIX] Tests after prevening SO duplicates

florian-dacosta committed:
Make exception rules unactived by default + add the no update flag to make it customizable

florian-dacosta committed:
[FIX] tests and ci

github-grap-bot committed:
[UPD] README.rst

github-grap-bot committed:
sale_import_base 14.0.1.1.0

sebastienbeau committed:
sale_import_base: refactor and migration to V16

sebastienbeau committed:
sale_import_base: use extendable_pydantic

florian-dacosta committed:
[FIX] Adapt schema to pydantic v2

florian-dacosta committed:
[FIX] tests

florian-dacosta committed:
[FIX] replace deprecated dict method by model_dump

florian-dacosta committed:
[FIX] Make main SaleOrder pydantic class extendable

florian-dacosta committed:
[FIX] warnings : add dummy security rule for sale.channel.importer + make sale.import.service.sale an Abstract Model

bealdav committed:
IMP sale_imp_base: remove product_unique_code

clementmbr committed:
[IMP] add crm_team_id and methods to ease inherit sale.channel.importer

clementmbr committed:
[IMP] search product related to its company_id

clementmbr committed:
[IMP] search product with no company_id

florian-dacosta committed:
[IMP] Add hook to allow to disable the invoice and delivery address merge and always have 2 distinct addresses
In case you want to update the delivery adresse after the creation of the sale order, it is easier to already have distinct addresses between delivery and invoicing.
Else one should create a new address, update the sale order, update the delivery order...
In the future, it  could become an option on channel or company instead of this hook that require an override to change the behavior

Commit:
[FIX] pre-commit clean for OCA

Commit:
[FIX] test mock and queue.job.chunk creation

florian-dacosta committed:
[REF] Refactoer queue_job_chunk and include it in sale_import_base
queue.job.chunk is renamed sale.import.payload

florian-dacosta committed:
[IMP] vacuum sale payload after 365 days by default + add chatter to log errors

Commit:
[IMP] add archive_addresses field

nayatec commented:
Thanks for this PR i appreciate how it simplifies the approach.
However, I have one concern: I noticed that channel_type (or any other parameter) can't be used any longer determine the 'processor' in the process function (via _get_processor). This change makes the module more specific than it should be, in my opinion, and reduces its flexibility. Could we reconsider maintaining the processor selection?

florian-dacosta commented:
> Thanks for this PR i appreciate how it simplifies the approach. However, I have one concern: I noticed that channel_type (or any other parameter) can't be used any longer determine the 'processor' in the process function (via _get_processor). This change makes the module more specific than it should be, in my opinion, and reduces its flexibility. Could we reconsider maintaining the processor selection?

The objective of this refactore, is change the old `queue.job.chunk` model. It has been designed to be very generic, but it made it confusing, and added a bit of complexity. And finally, there is not other module based on it doing something else that importing a sale order, which is the purpose of this module.

this `_get_processor` ws designed to be override in case you wanted to do something totally different than importing a `sale.order`. Is it your case ?
If you did use `queue.job.chunk` for both importing sale.order + doing something totally different (so with a totally different payload?), the idea would be to have a dedicated model for this totally other stuff. Si you can have a different menu, etc...

In an other hand, if you still import sale.order but some logic differs a bit depending on the `sale.channel`, then I don't see why you need this `_get_processor` logic. You could override `sale.channel.importer.run()` on any other submethods right ?
If a hook is needed somewhere for some use case, we could also add one.

Can you explains your use case or show me some related code so I understand better ?
(If necessary, I guess we'll put back some `_get_processor` logic of course, but I don't see why it is needed for now)

clementmbr commented on sale_import_base/models/sale_order.py:
```diff
@@ -0,0 +1,13 @@
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    si_amount_untaxed = fields.Float("(technical) Untaxed amount from import")
+    si_amount_tax = fields.Float("(technical) Tax amount from import")
+    si_amount_total = fields.Float("(technical) Total amount from import")
+    si_force_invoice_date = fields.Date("(technical) Forced invoice date")
+    si_force_invoice_number = fields.Char("(technical) Forced invoice number")
```
@florian-dacosta Does these "si_" prefixes come from ShopInvider? If it's the case, it could be better to rename them "imported_" or something explicit with no link with Shopinvider. Don't you think?

florian-dacosta replied:
Actually the si prefix is for sale import, not shopinvader!
So I don't think we really need to change it.

But, now that you point these fields, I see there is no logic at all linked to `si_force_invoice_date` and `si_force_invoice_number`.
I wonder if we should get rid of it. I it is needed in some case, then some logic should be implemented, and it should probably not belong to this base module IMO.

clementmbr replied:
I've never used the module in production, so you're more up to decide what's the best with these data.

Intuitively it looks like it can be useful to compare the imported amounts with the computed amounts during the SO creation with Odoo taxes (in case of cents differences for example).

But if it's not that useful to have these info on the SO, I imagine the user can always go back to the payload raw data in case he needs to make these comparatives.

florian-dacosta replied:
Well, the amounts fields are used already and I think it is good to keep.
But I am removing `si_force_invoice_date` and `si_force_invoice_number` since not used et not really standard.

clementmbr commented:
> In an other hand, if you still import sale.order but some logic differs a bit depending on the `sale.channel`, then I don't see why you need this `_get_processor` logic. You could override `sale.channel.importer.run()` on any other submethods right ?

Hi @florian-dacosta, entering the debate I agree with @nayatec that we do need something like the `get_processor()` (but I think it can be written in a more simple/useful way than in the current queue.job.chunk module).

The justification is that we need to be able to override the sale.channel.importer's methods in a specific way for each kind of sale.channel, even if all the sale.channel create sale.order :
For example : the Amazon channel will receive a specific payload which will need to be formatted in a specific way to fit the SaleOrder pydantic schemas, and we will also want to catch the delivery address by its external ID and update the existing SO instead of rising an Error.

So we will do 3 override of `_get_formatted_data()`, `_manage_existing_so()` and `_process_addresses()`... that should not be used by any other sale.channel used to create SaleOrder from any other source of payload (Ebay, local ecommerce...)

As far as I understand the problem, this means that we still need to create a new model with `_inherit=...` and `_name=...` for each kind of source of payload, am I right?

Practically we'll need to keep a `get_processor()` method inside the sale.import.payload's `process()`, but it can be used with a simple override adding a `if sale_channel_id.type == "amazon": return "sale.channel.importer.amazon"`, using the [`type` field already merged in v16.](https://github.com/OCA/sale-channel/blob/16.0/sale_channel/models/sale_channel.py#L14)

What do you think?

florian-dacosta commented:
@clementmbr Can you share some  code ?

>So we will do 3 override of _get_formatted_data(), _manage_existing_so() and _process_addresses()

And why can't you do these simple override with present code ?
```
def _get_formatted_data(self):
    if self.payload_id.sale_channel_id.channel_type == "amazon":
        return self._format_amazon_payload
    return super()._get_formatted_data()
```

Presently you have a new sale.channel.importer.amazon, inheriting the `sale.channel.importer` ?
I mean why not, but is it better that way ? I don't see how it helps?

If you both insists, I'll add back the `_get_processor` method, that will return `self.env["sale.channel.importer"].new({"payload_id": self.id})`

But I still don't really get why it is needed.

florian-dacosta commented:
Well, the current debat is more about the way to implement the specifities of a channel rather than if it is possible to do it (which it is).
I've updated the PR to allow to do as it was done, so everyone has the choice on how they implement their specificities.
I also added a hook `_get_validated_data` in the importer, that could allow to use a different pydantic schema.
Careful though, I think when using sale_import_rest, it is not so easy to manage different schemas.

clementmbr commented:
> @clementmbr Can you share some code ?

For example on this Amazon importer I need to override  5 different methods from sale.channel.importer : https://github.com/akretion/sale-import/blob/14-sale-import-amazon-backportv16/sale_import_amazon_vendor/models/sale_channel_importer_amazon_vendor.py

So it looks easier and cleaner to use a `_inherit=...` + `_name=...`
than using 5 `if self.payload_id.sale_channel_id.channel_type == "amazon":`

clementmbr commented on sale_import_base/models/sale_channel_importer.py:
```diff
@@ -0,0 +1,295 @@
+        return vals_addr_invoice == vals_addr_shipping
+
+    def _process_addresses(self, parent, address_invoice, address_shipping):
+        vals_addr_invoice = self._prepare_partner(address_invoice, parent.id, True)
+        vals_addr_shipping = self._prepare_partner(address_shipping, parent.id, True)
```
@florian-dacosta,

@nayatec suggested in a recent PR that I copied [in this last commit](https://github.com/akretion/sale-import/pull/95/commits/71e76d095d11cd25d1f09250331ba430868d8fdf), to add a `archive_addresses` boolean field instead of this always `True` parameter. Allowing the user to choose for each channel if he wants to archive the invoice/shipping addresses or not.

Do you think it's a good idea to implement in this sale_import_base module?

florian-dacosta replied:
Yes, definitly ! Where is the v16 commit ?

clementmbr replied:
It does not exist yet! :)
But I think you might be able to cherry pick this last one...

florian-dacosta replied:
Ok, I mis-understood, I'll cherry-pick it

clementmbr commented on sale_import_base/models/sale_import_payload.py:
```diff
@@ -0,0 +1,114 @@
+        self.enqueue_job()
+
+    def enqueue_job(self):
+        if DEBUG_MODE:
```
I remember here that @kevinkhao suggested to use a `if "pdb" in config.get("dev_mode"):` instead of hardcoded constant :
https://github.com/OCA/sale-channel/pull/20#discussion_r1616647264

What do you think @florian-dacosta ?

florian-dacosta replied:
I think this part is quite ugly, and I am wondering if it should not simply be removed.
But well, let's keep it for now, I replaced it with @kevinkhao suggestion, seems like a good idea to avoid changing the code during debug indeed.

florian-dacosta committed:
[FIX] Remove unused invoice data

clementmbr commented on sale_import_base/models/sale_channel_importer.py:
```diff
@@ -0,0 +1,296 @@
+        channel = self.payload_id.sale_channel_id
+        company_id = channel.company_id
+
+        product = self.env["product.product"].search(
+            [
+                ("default_code", "=", line_data["product_code"]),
+                ("product_tmpl_id.company_id", "=", company_id.id),
+            ]
+        )
+        if not product:
+            product = self.env["product.product"].search(
+                [
+                    ("default_code", "=", line_data["product_code"]),
+                    ("product_tmpl_id.company_id", "=", False),
+                ]
+            )
```
One more thing @florian-dacosta in a recent use case I needed to override the way the product is searched (searching on the barcode instead of searching on default_code), so I created this dedicated method on the v14 sale_import_base : https://github.com/akretion/sale-import/commit/7422ad73fbd4788ca54672cfcfa2d38ba3f66c33#diff-27e4483cdc08344f76a62e3859deaafd9e5ac59ddedb1a44d839c17b3b4c0b77R211

Do you think you can create this method here too?
Thanks!

florian-dacosta replied:
Make sense to search on barcode indeed. I wonder if this should not be configurable on channel or something.
Anyway, wouldn't it be enough to make a hook to get the domain [("default_code", "=", line_data["product_code"])] ?

It seems enough and easier to override.
Something like 
```
base_product_domain = self._get_product_domain(line_data)
product = self.env["product.product"].search(
          base_product_domain + 
              [("product_tmpl_id.company_id", "=", company_id.id)]
      )

def _get_product_domain(self, line_data):
    return [("default_code", "=", line_data["product_code"])] 

```

clementmbr replied:
Yes it might be better that way! I'll change my commit on the v14

florian-dacosta replied:
Hum right, I did not think about this.
Well maybe it is best to do as you did then. (We do not want multiple override for such a simple change...)
I'd only rename the `get_product_id` by `get_product` and same for `product_id` variable by `product`
if you don't mind, because I think it is more consistent to the way variable are defined globally in Odoo. (usually with `get_product_id` you would expect an id, and not a recordset) Well, not mandatory, I guess it is also a question of habit/preference, but it seems weird to me !

clementmbr commented on sale_import_base/models/sale_channel_importer.py:
```diff
@@ -0,0 +1,296 @@
+        if not product:
+            raise ValidationError(
+                _(
+                    "There is no active product with the Internal Reference %(code)s "
```
Following the idea the the product search domain can be override, we need to change the "Internal Reference" here by "code".

florian-dacosta replied:
Hum right, I did not think about this.
Well maybe it is best to do as you did then. (We do not want multiple override for such a simple change...)
I'd only rename the `get_product_id` by `get_product` and same for `product_id` variable by `product`
if you don't mind, because I think it is more consistent to the way variable are defined globally in Odoo. (usually with `get_product_id` you would expect an id, and not a recordset) Well, not mandatory, I guess it is also a question of habit/preference, but it seems weird to me !

clementmbr replied:
Thinking about it, I think there are pros and cons for both options :
-  the override on the _get_product_domain is easy to write, but we loose this information of the field's name during the raise
- the override on the _get_product has a more explicit raise.. But you have to rewrite the 2 raises in your override

clementmbr replied:
And thanks for the tips on the variable names ! :pray: 
I spent my whole life thinking it was mandatory to write records with a "_id" suffix, precisely in order to recognize that the variable is a record/recordset (I don't even remember who told me this or where I read this)

But today I discovered that I was wrong this whole time ! :fearful: 
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#variable-names

florian-dacosta replied:
> Thinking about it, I think there are pros and cons for both options :
> 
>     * the override on the _get_product_domain is easy to write, but we loose this information of the field's name during the raise
> 
>     * the override on the _get_product has a more explicit raise.. But you have to rewrite the 2 raises in your override

Well, we can do both. I just pushed a commit about this, let me know if it is ok for you.
I actually changed "Internal reference" by "code" in the error message.
 It makes it more generic in case of override, and I belive it stays totally understandable for the user. If some user check this kind of error, he must be aware of which field the product code is searched.
 Moreover, the second error was already talking about "code" and not "Internal reference".

clementmbr commented on sale_import_base/views/sale_import_payload.xml:
```diff
@@ -0,0 +1,92 @@
+                    <group>
+                        <group>
+                            <field name="create_date" />
+                            <field name="sale_channel_id" />
```
@florian-dacosta FYI another thing I found useful in practice is to add a `sale_order_id` field on the sale.import.payload to be filled with the created or existing SO, and a "N Payloads" head button on the sale.channel sending to the tree view of the related payloads.

florian-dacosta replied:
About the `sale_order_id`, it could be nice for filtering, but it is already doable searching for the ref in the `editable_data`, I did it this was until today and it did not bother me.
Linking existing SO may be not so simple because of the raise...

About a button on sale channel, you could do the same just filtering from the `sale.import.payload` menu because after this refactore you now have a field for `sale_channel_id` in the search view.
I guess it could be added anyway, I don't feel the need though.

Feel free to propose a PR with this kind of new stuff anyway. It should not block this refactore PR though.

florian-dacosta committed:
[IMP] Add hook for searching the sale order line product

florian-dacosta commented:
@nayatec @hparfr @clementmbr 
This one seems ready to me if you are able to review

sebastienbeau reviewed (approved):
LGTM (code review)

sebastienbeau commented:
/ocabot merge nobump
